File: 206_ReverseLinkedList.c
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
//69.89%
struct ListNode* reverseList(struct ListNode* head) {
	if(!head || !head -> next)
		return head;
    struct ListNode *pre = head, *curr = head -> next;
	pre -> next = NULL;
	struct ListNode *after = curr -> next;
	curr -> next = pre;
	pre = curr;
	curr = after;
	while(curr){
		after = curr -> next;
		curr -> next = pre;
		pre = curr;
		curr = after;
	}
	return pre;
}

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     struct ListNode *next;
 * };
 */
//2.69%
struct ListNode *reverseListHelper(struct ListNode **head){
	if(!((*head) -> next))
		return *head;
	struct ListNode *res = reverseListHelper(&((*head) -> next));
	(*head) -> next -> next = (*head);
	return res;
}
struct ListNode* reverseList(struct ListNode* head) {
	if(!head || !head -> next)
		return head;
	struct ListNode *res = reverseListHelper(&(head -> next));
	head -> next -> next = head;
	head -> next = NULL;
	return res;
}

File: 206_ReverseLinkedList.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None
# 54.72%
class Solution(object):
    def reverseList(self, head):
        """
        :type head: ListNode
        :rtype: ListNode
        """
        if not head or not head.next:
            return head
        pre, curr, after = head, head.next, head.next.next
        pre.next, curr.next = None, pre
        pre, curr = curr, after
        while curr:
            after = curr.next
            curr.next = pre
            pre = curr
            curr = after
        return pre
# Non-recursive		
class Solution:
# @param {ListNode} head
# @return {ListNode}
def reverseList(self, head):
    prev = None
    while head:
        curr = head
        head = head.next
        curr.next = prev
        prev = curr
    return prev


class Solution:
# @param {ListNode} head
# @return {ListNode}
def reverseList(self, head):
    return self._reverse(head)

def _reverse(self, node, prev=None):
    if not node:
        return prev
    n = node.next
    node.next = prev
    return self._reverse(n, node)	
        
        